content/advent-2017/go-grpc-beyond-basics.md

```diff
@@ -0,0 +1,500 @@
+
+### Securing our service
+
+If we look at the `grpc.NewServer` function definition (`func NewServer(opt ...ServerOption) *Server`) we will see that it is a [variadic function](https://blog.learngoprogramming.com/golang-variadic-funcs-how-to-patterns-369408f19085) that accepts a variable number of `grpc.ServerOption` values.
+
+To enable TLS for our service we need to use the `grpc.Creds` function which returns a `grpc.ServerOption` to send to the `grpc.NewServer` function.
+
+Let's look at the example.
+
+Service code:
+
+```go
+creds := credentials.NewTLS(&tls.Config{
+    // TLS config values here
+})
+
+serverOption := grpc.Creds(credentials.NewTLS(tlsConfig))
+server := grpc.NewServer(serverOption)
+```
+
+The code to create a `tls.Config` is standard Go. The real lines of code that's of interest are the following:
+
+``` go
+serverOption := grpc.Creds(credentials.NewTLS(tlsConfig))
+server := grpc.NewServer(serverOption)
+```
+
+The `credentials.NewTLS()` function construct a `credentials.TransportCredentials` value based on the `tls.Config` supplied.
+The `grpc.Creds()` funtional option takes the `credentials.TransportCredentials` value and sets credentials for server connections.
+
+Now the gRPC server will accept TLS connections.
+
+Let's turn our focus to enabling TLS on the client side.
+
+In the basic example we supply a `grpc.WithInsecure()` value to the `grpc.Dial` function. The `grpc.Insecure()` function returns a `DialOption` value which disables transport security for the client connection. By default, transport security is required so to disable transport security we need to set `WithInsecure`.
+But we want to enable TLS transport security. This is done with the `grpc.WithTransportCredentials()` function. Just like the `grpc.Creds()` function we used to enable transport security on the server side, the `grpc.WithTransportCredentials()` function also accepts a `credentials.TransportCredentials` but the difference is it returns a `DialOption` value and not a `ServerOption` value, and `grpc.Dial` function accepts `DialOption` values.
+
+Client code:
+
+``` go
+creds := credentials.NewTLS(&tls.Config{
+    //TLS Config values here
+})
+
+conn, err := grpc.Dial(*target, grpc.WithTransportCredentials(creds))
+if err != nil {
+    log.Fatalf("grpc.Dial err: %v", err)
+}
+```
+
+Now our service are enabled with TLS to encrypt data sent over the wire.
+
+There are many other options like message and buffer sizes and specifying a custom codec (something other than Protocol Buffer) available.To see what other options are available the API docs are your friend.
+
+For more server side options, see [ServerOption](https://godoc.org/google.golang.org/grpc#ServerOption).
+
+For more client side options, see [DialOption](https://godoc.org/google.golang.org/grpc#DialOption).
+
+**Tip:** `DialOption` values all have a **With** prefix, i.e grpc.**With**TransportCredentials
+
+*Fo more resources on enabling TLS for gRPC in Go:*
```

Comment:
"Fo" looks like a typo, should be "For".

Reply:
Fixed, thanks.